middleware/metrics.go

```diff
@@ -0,0 +1,51 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var metricsInfoKey = struct{}{}
+
+var (
+	RequestDurationOpts = prometheus.HistogramOpts{
+		Name:    "request_duration_seconds",
+		Help:    "Time (in seconds) spent serving this request.",
+		Buckets: prometheus.DefBuckets,
+	}
+)
+
+// MetricsInfo holds all of the metrics collected for a request.
+type MetricsInfo struct {
+	RequestDuration prometheus.Histogram
+	Start           time.Time
+}
+
+// GetMetrics retuns the metrics for this current context.
+func GetRequestMetrics(ctx context.Context) MetricsInfo {
+	metrics, ok := ctx.Value(metricsInfoKey).(*MetricsInfo)
+	if !ok || metrics == nil {
+		return MetricsInfo{
+			RequestDuration: nil,
+			Start:           time.Time{},
+		}
+	}
+	duration := time.Since(metrics.Start)
+	metrics.RequestDuration.Observe(duration.Seconds())
```

Comment:
Why is this here and not after the "next" call?

Reply:
You raise a good point. I was thinking that having a way to get the current duration of a running request would be useful for observability, but it's not really clear that it is. I think that requires a rethink, but moving the .Observe call to right after the next.Serve will handle what I was trying to do in a non-janky way.

My mind went off on a tangent about making a HistogramVec (collection of Histograms) that we can access via this call. Something to think about.